query/src/sql/planner/plans/expression.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+
+use crate::sql::optimizer::PhysicalProperty;
+use crate::sql::optimizer::RelationalProperty;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::BasePlan;
+use crate::sql::plans::LogicalPlan;
+use crate::sql::plans::PhysicalPlan;
+use crate::sql::plans::PlanType;
+use crate::sql::ScalarExprRef;
+
+#[derive(Clone)]
+pub struct ExpressionPlan {
```

Comment:
This is the same as `Project` operator, you should use that instead.

By the way, I don't think there is requirement for this. Aggregate plan should have produced the aggregate functions in its output, which can be referenced by the following `HAVING` and `SELECT` clauses.

Reply:
As leiysky suggested, will we do it in this PR or another? @xudong963

Reply:
Sorry for the late reply, I took a day off yesterday.

I use `ExpressionPlan` to process scalar expression in group by, such as group by a+1. Its role is to bridge the schema of scan plan and aggregate plan. I didn't use the direct project operator because it's a bit strange to have two project operators in the whole plan tree, and there is no need to record column index in Expression Plan.

Reply:
> Sorry for the late reply, I took a day off yesterday.
> 
> I use `ExpressionPlan` to process scalar expression in group by, such as group by a+1. Its role is to bridge the schema of scan plan and aggregate plan. I didn't use the direct project operator because it's a bit strange to have two project operators in the whole plan tree, and there is no need to record column index in Expression Plan.


I got you.

The point is, every expression can be evaluated on the fly with a context, so we don't need a `ExpressionPlan` to materialize it in the context.

A special case is projection. In projection, an expression can be assigned to a variable through `expr AS "var"`, which makes it can be referenced by other expressions. Canonically, all the variable `"var"` are directly replaced by `expr`. But we don't like that, it's meaningless.

Our solution is, for the expressions(aliased expression in projection or aggregate function produced by `GROUP BY`) that can be explicitly referenced by other expressions, we will store it in the context as a column. In particular, we give the `Scalar` a unique identifer(i.e. `column_index`) and store it in `BindContext` as a `ColumnBinding`.

For the rest expressions, since most of them are not reuseable, we just store them inside the `Plan`s where they are needed.

In your case, there is no need to pre-evaluate some expressions for `GROUP BY`, since they are already involved in the `AggregatePlan`.

After all, there is no `schema` in the planning phase. It's all about logical representation of relational algebra.

If you are confused by `schema`, I'm betting you encountered a execution issue. Try to fix it in `PipelineBuilder`.